runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,226 @@
+package org.corfudb.runtime.collections;
+
+import com.google.common.reflect.TypeToken;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.object.transactions.TransactionType;
+import org.corfudb.runtime.object.transactions.TransactionalContext;
+import org.corfudb.runtime.view.CorfuGuidGenerator;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+
+/**
+ * Persisted Queue supported by CorfuDB using distributed State Machine Replication.
+ * Entries enqueued are backed by a CorfuTable with a LinkedHashMap where each entry
+ * is mapped to a unique generated id that is returned upon successful <b>enqueue()</b>.
+ * <b>entryList()</b> returns the enqueued entries along with their special ids which can
+ * not only identify the element for removal but also has a global ordering defined by
+ * the order in which their <b>enqueue()</b> operations materialized.
+ * Instead of a dequeue() this Queue supports a <b>remove()</b> which accepts the id of the element.
+ * Entries cannot be modified in-place (or will lose ordering) but can be removed from anywhere
+ * from the persisted queue.
+ *
+ * Created by hisundar on 5/8/19.
+ *
+ * @param <E>   Type of the entry to be enqueued into the persisted queue
+ */
+@Slf4j
+public class CorfuQueue<E> {
+    /**
+     * The main CorfuTable which contains the primary key-value mappings.
+     */
+    private final CorfuTable<Long, E> corfuTable;
+    private final CorfuRuntime runtime;
+    private final CorfuGuidGenerator guidGenerator;
+
+    public CorfuQueue(CorfuRuntime runtime, String streamName) {
+        this.runtime = runtime;
+        corfuTable = runtime.getObjectsView().build()
```

Comment:
can you please pass CorfuTable directly instead of CorfuRuntime

Reply:
CorfuTable is internal to the implementation of this data structure. Runtime is needed not just for building the CorfuTable but also for transactions and snapshot versions. Please suggest an alternative if not for runtime?